### PR TITLE
accounts/mananger, internal/ethapi/api: Add new function AllAccounts …

### PR DESCRIPTION
**Pull request by rociiu**
…on account manager to remove the duplication code on getting all wallets accounts

Refactor: remove the duplication code on getting all account addresses, extract the logic into account manager .

### CODE REVIEW

**Comment by holiman** [comment]
Looks generally good to me, but the new code is using `am.lock.RLock()` which wasn't used before. Why is that needed now -- was it just a miss that it wasn't present earlier?

**Comment by rociiu** [comment]
@holiman thanks for review. In the old way, the lock/unlock happened in `s.am.Wallets()`, since we're moving the code inside the `s.am`, the lock/unlock is make more sense to handle directly inside the function. Otherwise we have to call `am.Wallets()` to retrieve the wallets which cause extra copy of the wallets data.

File: accounts/manager.go [hunk 1]
```diff
@@ -147,6 +148,20 @@ func (am *Manager) Wallet(url string) (Wallet, error) {
 	return nil, ErrUnknownWallet
 }
 
+// AllAccounts returns all account addresses of all wallets within the account manager
```
Comment (karalabe) [inline accounts/manager.go]:
```suggestion
// Accounts returns all account addresses of all wallets within the account manager
```

File: accounts/manager.go [hunk 2]
```diff
@@ -147,6 +148,20 @@ func (am *Manager) Wallet(url string) (Wallet, error) {
...
+func (am *Manager) AllAccounts() []common.Address {
```
Comment (karalabe) [inline accounts/manager.go]:
```suggestion
func (am *Manager) Accounts() []common.Address {
```

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -190,13 +190,7 @@ func NewPublicAccountAPI(am *accounts.Manager) *PublicAccountAPI {
...
 // Accounts returns the collection of accounts this node manages
 func (s *PublicAccountAPI) Accounts() []common.Address {
-	addresses := make([]common.Address, 0) // return [] instead of nil if empty
-	for _, wallet := range s.am.Wallets() {
-		for _, account := range wallet.Accounts() {
-			addresses = append(addresses, account.Address)
-		}
-	}
-	return addresses
+	return s.am.AllAccounts()
```
Comment (karalabe) [inline internal/ethapi/api.go]:
```suggestion
	return s.am.Accounts()
```

File: internal/ethapi/api.go [hunk 4]
```diff
@@ -219,13 +213,7 @@ func NewPrivateAccountAPI(b Backend, nonceLock *AddrLocker) *PrivateAccountAPI {
...
 // ListAccounts will return a list of addresses for accounts this node manages.
 func (s *PrivateAccountAPI) ListAccounts() []common.Address {
-	addresses := make([]common.Address, 0) // return [] instead of nil if empty
-	for _, wallet := range s.am.Wallets() {
-		for _, account := range wallet.Accounts() {
-			addresses = append(addresses, account.Address)
-		}
-	}
-	return addresses
+	return s.am.AllAccounts()
```
Comment (karalabe) [inline internal/ethapi/api.go]:
```suggestion
	return s.am.Accounts()
```

**Pull request merged**